### PR TITLE
Remove usage of numpy, action list DSL, and various small refactors

### PR DESCRIPTION
**Pull request by rht**
- [x] wip ensure the examples are working correctly just as before

### CODE REVIEW

**Comment by DavoudTaghawiNejad** [comment]
I don't understand the purpose of this:
```

     def add_agents(self, messages, round):
 +        if len(messages) == 0:
 +            return
          for _, _, (AgentClass, group_name, parameters, agent_parameters) in messages:
              id = self.num_of_agents_in_group[group_name]
              self.num_of_agents_in_group[group_name] += 1
```

**Comment by DavoudTaghawiNejad** [comment]
```
+            self.rounds -= 1
+        if self.rounds == 0:
+            self._finalize()

```
wouldn't that be nicer:
```

+ if self.round == self.rounds:
+    self._finalize()
```

**Comment by DavoudTaghawiNejad** [comment]
I don't understand this (we talked about that):
```

+        messages[('_simulation', 0)] = []
 +        messages[('_simulation', 0.5)] = []
 +        self.sim._agents_to_add.extend(messages.pop(('_simulation', 0), []))
 +        self.sim._agents_to_delete.extend(messages.pop(('_simulation', 0.5), []))
```

Does it not delete the add/remove messages?

File: abce/__init__.py [hunk 1]
```diff
@@ -717,8 +591,12 @@ def build_agents(self, AgentClass, group_name, number=None, parameters={}, agent
 
             self.family_list[group_name].append(family)
             self.num_of_agents_in_group[group_name] = num_agents_this_group
+        return Group(self, self.family_list[group_name], group_name)
 
-    def add_agents(self, messages, round):
+    def add_agents(self, round):
+        messages = self._agents_to_add
+        if len(messages) == 0:
+            return
```
Comment (rht) [inline abce/__init__.py]: > I don't understand the purpose of this:

Comment moved here for easier multithreaded reply.

File: abce/__init__.py [hunk 2]
```diff
@@ -547,80 +453,40 @@ def execute_internal_parallel(self, command):
...
+        if self.round == self._start_round:  # the simulation has just started
+            self._prepare()
+        while self.rounds > 0:
+            if self.round > self._start_round:
+                self._finalize_prev_round(self.round - 1)
+            self._prepare_round(self.round)
+            yield self.round
+            self.round += 1
+            self.rounds -= 1
+        if self.rounds == 0:
```
Comment (rht) [inline abce/__init__.py]: > wouldn't that be nicer:
> `if self.round == self.rounds:`

File: abce/group.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    def __add__(self, g):
+        return Group(self.sim, self.groups + g.groups)
+
+    def execute_serial(self, command):
+        messages = self.sim.messagess
+        for family in self.group:
+            messages[family.name()] = family.execute(command, messages[family.name()])
+        messages[('_simulation', 0)] = []
+        messages[('_simulation', 0.5)] = []
+        self.sim._agents_to_add.extend(messages.pop(('_simulation', 0), []))
```
Comment (rht) [inline abce/group.py]: If you trace these lines to their current location in master, they are pretty much equivalent to this, and so are made more explicit this way.

File: abce/__init__.py (hunk 2 above)
Comment (rht) [inline abce/__init__.py]: ok

File: abce/__init__.py (hunk 1 above)
Comment (rht) [inline abce/__init__.py]: This is basically https://github.com/DavoudTaghawiNejad/abce/blob/d15ceaf956db98639fd2556db15e0af65eac7aab/abce/__init__.py#L604 moved here.

Comment (DavoudTaghawiNejad) [inline abce/__init__.py]: yes but why do you test that len(mhessages) is 0? If the list is empty the the following for loop would anyway not do anything?

File: abce/group.py (hunk 3 above)
Comment (DavoudTaghawiNejad) [inline abce/group.py]: So its potentially wrong, but you just keep it? If you confirm I will open an issue.

File: abce/__init__.py (hunk 1 above)
Comment (rht) [inline abce/__init__.py]: If the list is empty then there are no agents to add.

File: abce/group.py (hunk 3 above)
Comment (rht) [inline abce/group.py]: I see now. L14 actually does include a message with the key ('_simulation', 0) and ('_simulation', 0.5), so L15 and L16 are possibly no longer necessary.

Comment (rht) [inline abce/group.py]: Actually, this should be left for another pr.

**Comment by rht** [comment]
@DavoudTaghawiNejad all the examples have been fixed -- this is ready to merge.

File: abce/__init__.py (hunk 1 above)
Comment (DavoudTaghawiNejad) [inline abce/__init__.py]: Yes, but its redundant, if the list is empty the for loop loops over nothing. Am I missing something?
